### PR TITLE
FEATURE: Initial functionality

### PR DESCRIPTION
**Pull request by stankowalczyk**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,4 +1,533 @@
...
+   */
+  getListsRelevantToStatusUpdate(trelloBoardId, time) {
+    return this.getAllListsOnBoard(trelloBoardId)
+    .then(lists => {
+      let completedTaskList     = lists.find(list => TrelloExtractor.isCompletedList(list, time));
+      let thisWeeksUpcomingList = lists.find(list => TrelloExtractor.isThisWeekUpcomingList(list, time));
+      let nextWeeksUpcomingList = lists.find(list => TrelloExtractor.isNextWeekUpcomingList(list, time));
+      let currentTaskLists      = TrelloExtractor.isCurrentTaskList(lists);
+
+      let promiseArray = [
```
Comment (lukehorvat) [inline lib/index.js]:
```js
return Promise.all([
  this.getCardsFromTrelloList(completedTaskList),
  this.getCardsFromTrelloList(thisWeeksUpcomingList),
  this.getCardsFromTrelloList(nextWeeksUpcomingList),
  ...currentTaskLists.map(list => this.getCardsFromTrelloList(list)))
])
```

File: lib/index.js [hunk 2]
```diff
@@ -1,4 +1,533 @@
...
+        this.getCardsFromTrelloList(thisWeeksUpcomingList),
+        this.getCardsFromTrelloList(nextWeeksUpcomingList)
+      ];
+
+      currentTaskLists.forEach(list => {
+        promiseArray.push(this.getCardsFromTrelloList(list));
+      });
+
+      return Promise.all(promiseArray)
+      .then(([completeCards,
```
Comment (lukehorvat) [inline lib/index.js]: Use Bluebird's `.spread` instead of `.then` here.

File: lib/index.js [hunk 3]
```diff
@@ -1,4 +1,533 @@
...
+    return this.getBoardDetails(trelloBoardId)
+    .then(boardResult => {
+      result = boardResult;
+      return this.getListsRelevantToStatusUpdate(trelloBoardId, comparisonTime);
+    }).then(cardResults => {
+      result.completedTasks = cardResults.complete;
+      result.currentTasks = cardResults.current;
+      result.plannedTasks = cardResults.planned;
+
+      return Promise.resolve(result);
```
Comment (lukehorvat) [inline lib/index.js]: Don't need `Promise.resolve`.

File: lib/index.js (hunk 1 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 2 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js [hunk 4]
```diff
@@ -1,4 +1,533 @@
...
+
+  /**
+   * Gets the date and time a specific Trello Board was modified
+   * @param {string} boardId
+   * @return {Bluebird Promise} A promise containing a moment object representing the date and time a board was last updated.
+   */
+  getBoardLastModifiedDateTime(boardId) {
+    return this.trelloClient.getAsync(`/1/boards/${boardId}?fields=dateLastActivity`)
+    .then(board => {
+      return Promise.resolve(moment(board.dateLastActivity));
```
Comment (lukehorvat) [inline lib/index.js]: `Promise.resolve` not needed here either. it's only really needed when you want to start a *new* Promise chain.

File: lib/index.js [hunk 5]
```diff
@@ -1,4 +1,533 @@
...
+        return firstDate.isAfter(comparisonTime, 'day');
+      }
+    }
+    return false;
+  }
+
+  static isCurrentTaskList(listArray) {
+    let result = [];
+
+    listArray.forEach(list => {
```
Comment (lukehorvat) [inline lib/index.js]: Try and eliminate temporary variables when possible. it's just additional state that the programmer has to juggle in their head when reading the code. You could use `listArray.filter` here  think.

File: lib/index.js (hunk 3 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 4 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js [hunk 6]
```diff
@@ -1,4 +1,533 @@
...
+   */
+  static formatCardsFromList(cards) {
+    let result = [];
+
+    if (!cards || cards.length === 0) {
+      return result;
+    }
+
+    // Format each card, also format idMembers into the assignedPeople format
+    cards.forEach(card => {
```
Comment (lukehorvat) [inline lib/index.js]: Use `cards.map` to eliminate `result`.

File: lib/index.js (hunk 5 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js [hunk 7]
```diff
@@ -1,4 +1,533 @@
...
+  }
+
+  /**
+   * Returns Trello Member object from API
+   * @param {string} memberId Trello user id, or username
+   * @param {Bluebird Promise} A promise holding member details
+   */
+  getMemberDetails(memberId) {
+    if (!memberId) {
+      return Promise.resolve();
```
Comment (lukehorvat) [inline lib/index.js]: Shouldn't it throw an error here?

File: lib/index.js [hunk 8]
```diff
@@ -1,4 +1,533 @@
...
+    } else if (!card.idMembers || card.idMembers.length === 0) {
+      // If the card doesn't have people assigned to it, return the card as is
+      return Promise.resolve(card);
+    }
+
+    let promiseArray = [];
+
+    // Get member details for the card
+    card.idMembers.forEach(memberId => {
```
Comment (lukehorvat) [inline lib/index.js]: `card.idMembers.map`

File: lib/index.js (hunk 6 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js [hunk 9]
```diff
@@ -1,4 +1,533 @@
...
+  /**
+   * Gets board details from Trello, returns an object representing the Trello Board wrapped in a promise.
+   * @param {string} trelloBoardId
+   * @return {Bluebird Promise} A promise which holds a representation of a Trello Board.
+   */
+  getBoardDetails(trelloBoardId) {
+    let result = {};
+    return this.trelloClient.getAsync(`/1/boards/${trelloBoardId}?fields=name,desc,url,dateLastActivity`)
+    .then(board => {
+      result.name = board.name;
```
Comment (lukehorvat) [inline lib/index.js]: .then(board => {
  return {
    name: ...,
    description: ...
  };
})

File: lib/index.js [hunk 10]
```diff
@@ -1,4 +1,526 @@
...
+   */
+  getCurrentTrelloStatus(trelloBoardId, time) {
+    let result = {};
+    let comparisonTime = moment();
+
+    if (!!time && moment.isMoment(time)) {
+      comparisonTime = time;
+    }
+
+    return this.getBoardDetails(trelloBoardId)
```
Comment (lukehorvat) [inline lib/index.js]:
```js
return Promise.all([
  this.getBoardDetails(trelloBoardId),
  this.getListsRelevantToStatusUpdate(trelloBoardId, comparisonTime)
]).spread((boardResult, cardResults) => {
  return Object.assign({}, boardResult, {
    completedTasks: cardResults.complete,
    currentTasks: cardResults.current,
    plannedTasks: cardResults.planned
  });
});
```

File: lib/index.js [hunk 11]
```diff
@@ -1,4 +1,526 @@
...
+    return result;
+  }
+
+  /**
+   * Returns an object based on an array of completed cards, and the list they were sourced from
+   * @param {Array} arrayOfCards An array of cards
+   * @param {Object} sourceList The Trello List the cards were sourced from
+   * @return {Object} An object containing the cards and name of list
+   */
+  static formatCompleteTasks(arrayOfCards, sourceList) {
```
Comment (lukehorvat) [inline lib/index.js]: Should be functionally equivalent:

```js
static formatCompleteTasks(arrayOfCards = [], sourceList) {
  let tasks = TrelloExtractor.formatCardsFromList(arrayOfCards);
  let sources = [].concat(arrayOfCards.length > 0 ? sourceList.name : []);
  return { tasks, sources };
}
```

File: lib/index.js [hunk 12]
```diff
@@ -1,4 +1,526 @@
...
+      };
+    }
+  }
+
+  /**
+   * Formats an array of Trello Member objects into the platform agnostic "Assigned People" object structure
+   * @param {Array of Trello Member Objects} memberDetailsArray
+   * @return {Array of Assigned People Objects}
+   */
+  static formatMemberDetailsArray(memberDetailsArray) {
```
Comment (lukehorvat) [inline lib/index.js]:
```js
static formatMemberDetailsArray(memberDetailsArray = []) {
  return memberDetailsArray.filter(memberDetails => (
    !!TrelloExtractor.formatMemberDetails(memberDetails)
  });
}
```

File: lib/index.js [hunk 13]
```diff
@@ -1,4 +1,526 @@
...
+      return Promise.resolve();
+    } else if (!card.idMembers || card.idMembers.length === 0) {
+      // If the card doesn't have people assigned to it, return the card as is
+      return Promise.resolve(card);
+    }
+
+    let promiseArray = [];
+
+    // Get member details for the card
+    card.idMembers.map(memberId => {
```
Comment (lukehorvat) [inline lib/index.js]: Wasn't quite what i meant.

Comment (lukehorvat) [inline lib/index.js]:
```js
return Promise.all(
  card.idMembers.map(memberId => this.getMemberDetails(memberId))
).then(members => {
```

File: lib/index.js [hunk 14]
```diff
@@ -1,4 +1,526 @@
...
+
+  /**
+   * Returns an object based on an array of current cards, and the list they were sourced from
+   * @param {Array} arrayOfCardArrays An array containing arrays of cards
+   * @param {Array} arrayOfSourceLists An array containing lists from which the cards were sourced from
+   * @return {Object} An object containing the cards and name of list
+   */
+  static formatCurrentTasks(arrayOfCardArrays, arrayOfSourceLists) {
+    // arrayOfCardArrays is an array of nested arrays. Remove the layering
+    let cards = [];
```
Comment (lukehorvat) [inline lib/index.js]: Flatten multiple arrays into one array:

```js
let cards = arrayOfCardArrays.reduce((arr, arrayOfCards) => arr.concat(arrayOfCards), []);
```

File: lib/index.js [hunk 15]
```diff
@@ -1,4 +1,526 @@
...
+  }
+
+  /**
+   * Returns a list of cards formatted into a more generic representation of a task.
+   * @param {Array} cards Array of Trello cards
+   * @return {Array} An array of task objects
+   */
+  static formatCardsFromList(cards) {
+    if (!cards || cards.length === 0) {
+      return result;
```
Comment (lukehorvat) [inline lib/index.js]: `result`?

Comment (lukehorvat) [inline lib/index.js]: i would make the `cards` parameter have a default value of `[]`, then you don't even need the if statement at all.

**Comment by apj** [comment]
@stankowalczyk Agree with Luke's suggested changes. Overall, the code is looking much cleaner, but I think this one file is doing far too much. I discussed with Luke today and the focus for this PR should be to wrap up the quality changes outlined with a view of having this functionally ready for generated updates.

Beyond this iteration, I'd like to see the filtering of cards (i.e. DSTIL's conventions for grouping cards) decoupled from the interaction with the Trello API and output of tasks from cards. The way I see this working is to extract filtering/grouping into its own class, which would then be injected into an update extractor. The grouping class would expose methods for determining whether lists are Planned, Current or Completed, and the extractor would simply call those methods to determine which bucket the list/cards go into. While there's not an immediate goal of having the ability to inject alternative groupers, this change would make this package much easier to understand.

File: lib/index.js [hunk 16]
```diff
@@ -1,4 +1,526 @@
...
+    return false;
+  }
+
+  /**
+   * Determines if the list we are examining is 'Next Weeks' list
+   * @param {Trello List JSON} inputList
+   * @param {Moment instance} time The time which a status update should be generated for (if left blank, current time is used).
+   * @return {Boolean} True if the list name matches 'Next Weeks' list name style
+   */
+  static isNextWeekUpcomingList(inputList, time) {
```
Comment (apj) [inline lib/index.js]: @stankowalczyk Am I right in saying that this function, `isCompletedList `, and `isThisWeekUpcomingList` are the same aside from the comparison between `comparisonTime` and `firstDate` and/or `secondDate`? There should be a pretty cheap win in cutting down on duplication by refactoring this. We can discuss a good way to do this tomorrow if you like.

File: lib/index.js (hunk 9 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 10 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 11 above)
Comment (stankowalczyk) [inline lib/index.js]: Adapted slightly, cause I realised my if statement logic was flawed. Done.

File: lib/index.js (hunk 12 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 13 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 14 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 15 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

**Comment by stankowalczyk** [comment]
@apj Agreed, I can move that functionality out when the CLI is being built.

File: lib/index.js (hunk 16 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 8 above)
Comment (stankowalczyk) [inline lib/index.js]: Outdated - done.

File: lib/index.js (hunk 7 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: test/index.js [hunk 17]
```diff
@@ -1,8 +1,175 @@
-import should from "should";
-import trelloExtractor from "../lib";
+import fs from 'fs';
+import moment from 'moment';
+import nock from 'nock';
+import path from 'path';
+import Promise from 'bluebird';
+import should from 'should';
+import trelloExtractor from '../lib';
```
Comment (lukehorvat) [inline test/index.js]: Given it's a class, can you capitalise the T in `trelloExtractor`?

File: lib/index.js [hunk 18]
```diff
@@ -1,4 +1,434 @@
...
+
+  /**
+   * Gets all cards from a Trello list, in the form of a promise.
+   * @param {Promisified Trello Object} trello
+   * @param {Trello List JSON} list
+   * @return {Bluebird Promise} A promise which returns an array of cards
+   */
+  getCardsFromTrelloList(list) {
+    if (!list) {
+      return Promise.resolve();
```
Comment (lukehorvat) [inline lib/index.js]: Is this ever a possibility? When is `list` undefined/null?

Comment (lukehorvat) [inline lib/index.js]: Furthermore, it might be better to throw an error if this is unexpected behaviour.

File: lib/index.js [hunk 19]
```diff
@@ -1,4 +1,434 @@
...
+
+  /**
+   * Retrieves and replaces idMembers with Trello user details
+   * @param {Trello Card} card
+   * @param {Bluebird Promise} A promise holding card with updated idMembers field
+   */
+  resolveMemberDetailsFromCard(card) {
+    // If the card isn't defined - return
+    if (!card) {
+      return Promise.resolve();
```
Comment (lukehorvat) [inline lib/index.js]: Throw error?

File: lib/index.js [hunk 20]
```diff
@@ -1,4 +1,434 @@
...
+    let promiseArray = [];
+
+    // Get member details for the card
+    return Promise.all(
+      card.idMembers.map(memberId => this.getMemberDetails(memberId))
+    ).then(members => {
+      // Replace idMembers with real data (replacing the original data which was just ID's)
+      card.idMembers = members;
+
+      return Promise.resolve(card);
```
Comment (lukehorvat) [inline lib/index.js]: Don't need this `Promise.resolve`.

File: lib/index.js [hunk 21]
```diff
@@ -1,4 +1,434 @@
...
+  getCardsFromTrelloList(list) {
+    if (!list) {
+      return Promise.resolve();
+    }
+
+    return this.trelloClient.getAsync(`/1/lists/${list.id}/cards?fields=dateLastActivity,desc,idMembers,name,url`)
+    .then(cards => {
+      // If we have no cards, return an empty array
+      if (!cards || cards.length === 0) {
+        return Promise.resolve([]);
```
Comment (lukehorvat) [inline lib/index.js]: Don't need `Promise.resolve`.

File: lib/index.js [hunk 22]
```diff
@@ -1,4 +1,434 @@
...
+        this.getCardsFromTrelloList(completedTaskList),
+        this.getCardsFromTrelloList(thisWeeksUpcomingList),
+        this.getCardsFromTrelloList(nextWeeksUpcomingList),
+        ...currentTaskLists.map(list => this.getCardsFromTrelloList(list))
+      ]).spread((completeCards,
+        thisWeeksUpcomingCards,
+        nextWeeksUpcomingCards,
+        ...currentCards) => {
+
+        return Promise.resolve({
```
Comment (lukehorvat) [inline lib/index.js]: Don't need `Promise.resolve`.

File: lib/index.js [hunk 23]
```diff
@@ -1,4 +1,434 @@
...
+   * @param {Promisified Trello Object} trello
+   * @param {Trello List JSON} list
+   * @return {Bluebird Promise} A promise which returns an array of cards
+   */
+  getCardsFromTrelloList(list) {
+    if (!list) {
+      return Promise.resolve();
+    }
+
+    return this.trelloClient.getAsync(`/1/lists/${list.id}/cards?fields=dateLastActivity,desc,idMembers,name,url`)
```
Comment (lukehorvat) [inline lib/index.js]: Can be simplified:

```js
  return this.trelloClient.getAsync(`/1/lists/${list.id}/cards?fields=dateLastActivity,desc,idMembers,name,url`)
  .then((cards = []) => {
    return Promise.resolve(cards.filter(card => card.idMembers && card.idMembers.length > 0))
    .each(card => this.resolveMemberDetailsFromCard(card))
    .then(() => cards);
  });
```

File: lib/index.js [hunk 24]
```diff
@@ -1,4 +1,434 @@
...
+  resolveMemberDetailsFromCard(card) {
+    // If the card isn't defined - return
+    if (!card) {
+      return Promise.resolve();
+    } else if (!card.idMembers || card.idMembers.length === 0) {
+      // If the card doesn't have people assigned to it, return the card as is
+      return Promise.resolve(card);
+    }
+
+    let promiseArray = [];
```
Comment (lukehorvat) [inline lib/index.js]: Not used.

File: lib/index.js [hunk 25]
```diff
@@ -1,4 +1,434 @@
...
+
+    return this.trelloClient.getAsync(`/1/members/${memberId}`);
+  }
+
+  /**
+   * Retrieves and replaces idMembers with Trello user details
+   * @param {Trello Card} card
+   * @param {Bluebird Promise} A promise holding card with updated idMembers field
+   */
+  resolveMemberDetailsFromCard(card) {
```
Comment (lukehorvat) [inline lib/index.js]:
```js
resolveMemberDetailsFromCard(card) {
  if (!card) throw new Error("Card not specified.");

  return Promise.resolve(card).tap(card => {
    if (card.idMembers && card.idMembers.length > 0) {
      return Promise.resolve(card.idMembers)
      .map(memberId => this.getMemberDetails(memberId))
      .then(members => card.idMembers = members)
    }
  })
}
```

File: test/index.js (hunk 17 above)
Comment (stankowalczyk) [inline test/index.js]: Done.

File: lib/index.js (hunk 18 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 21 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 22 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 23 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

File: lib/index.js (hunk 25 above)
Comment (stankowalczyk) [inline lib/index.js]: Done.

**Pull request merged**